Don't Discard the create_channel for a suddenly disconnected peer

shaavan opened the pull request:
Resolves #2096 

- The PR allows the creation of the channel when the peer has disconnected during the creation process and allows rebroadcasting of the SendOpenChannel msg when they reconnect.
- If they are not able to reconnect in the two-timer tick time (around 2 mins), we force close the channel on our side.

shaavan commented:
This PR makes the following interpretation of the issue and follows the solution accordingly:

1. Don't allow channel creation if the peer is already disconnected.
2. But if the peer is connected, before calling the "create_channel" and disconnects at the exact time when the function is called. Allow channel creation, and resend the _open_channel message_ again when it reconnects.
3. Fail the creation after a few timer ticks if the peer fails to connect within the time.

If in case, my interpretation of the problem has been erroneous, do let me and I shall be glad to correct it! :)

shaavan commented:
Also, this PR has been set to draft because the tests are incomplete and only partially test the added code.
Any help or suggestion in making the tests work is much welcome!

shaavan commented:
> What's the status here. You have it marked draft, do you want feedback? What kind of feedback/review is this ready for?

Hi, @TheBlueMatt
I have completed the implementation in the PR, but the problem I am facing is writing up the test for it.

So, I am facing trouble preparing a test for when the peer is reconnected on time, and hence the open channel message is sent to it because it conflicts with how the rest of the test codebase is set up.

So I wanted to get a general _Approach ACK_, before I go about hacking in the test to make them work.

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -721,6 +721,9 @@ pub(super) struct PeerState<SP: Deref> where SP::Target: SignerProvider {
 	pub(super) inbound_channel_request_by_id: HashMap<ChannelId, InboundChannelRequest>,
+	///
+	/// Outbound channel unsent by us.
+	pub(super) outbound_channel_unsent_by_id: HashMap<ChannelId, OutboundChannelRequest>,
```
Let me try making the struct simpler making it store the list of channels for which we have to do the rebroadcast of `SendOpenChannel` and use reconstruction of information later as much as possible.

shaavan replied:
Addressed in [929deda](https://github.com/lightningdevkit/rust-lightning/commit/929deda1facf5f57df769b7c597e59c93c76540e)

Thanks for the suggestion!

shaavan commented:
Updated from [pr2725.01](https://github.com/shaavan/rust-lightning/commits/pr2725.01) -> [pr2725.02](https://github.com/shaavan/rust-lightning/commits/pr2725.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.01..pr2725.02))
Addressed @TheBlueMatt comment

Changes:

1. Rebased on Main.
2. Used data in `Channel` to reconstruct the SendOpenChannel message when rebroadcasting.
3. Clean up the faulty test.

Note:
1. The PR has been moved from "Draft" to "Ready for Review"
2. The tests were becoming cumbersome and hacky, so I have temporarily cleaned them up.
3. I am looking for a general Approach ACK on this PR before working on creating a test for this approach.

shaavan commented:
Updated from [pr2725.02](https://github.com/shaavan/rust-lightning/commits/pr2725.02) -> [pr2725.03](https://github.com/shaavan/rust-lightning/commits/pr2725.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.02..pr2725.03))

Changes:
- Updated the `peer_disconnected` function to keep the unnotified channel around in case of sudden disconnection.
- Also updated the `peer_connected` code to handle this change in behavior.

Thanks for the suggestion, @wpaulino!

shaavan commented:
Updated from [pr2725.03](https://github.com/shaavan/rust-lightning/commits/pr2725.03) -> [pr2725.04](https://github.com/shaavan/rust-lightning/commits/pr2725.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.03..pr2725.04))
Addressed @TheBlueMatt suggestion

Changes:
1. Rebased on Main.
2. Revamped the approach. Instead of sending the SendOpenChannel message later, we track if we have received an accepted channel message from them. This simplifies the approach and is in line with the current code architecture.
3. Added test to verify this new behavior.

Thank you very much, @TheBlueMatt, for this new idea to solve this problem.

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -742,6 +742,9 @@ pub(super) struct PeerState<SP: Deref> where SP::Target: SignerProvider {
 	pub(super) inbound_channel_request_by_id: HashMap<ChannelId, InboundChannelRequest>,
+	///
+	/// Channels that were created but not notified about by sending SendOpenChannel Message
+	pub(super) unnotified_outbound_channel: HashMap<ChannelId, OutboundChannelTimer>,
```
This approach is a much cleaner solution to the problem, and I have implemented this in [10fabf4](https://github.com/lightningdevkit/rust-lightning/commit/10fabf4266a33bf3ee7684eea54bdf55a328846a)

Thank you very much for the suggestion!

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8985,20 +9031,47 @@ where
+					}
+					else if peer_state.unnotified_outbound_channel.contains_key(chan_id) {
+						// That means the channel is notified about just now.
+						continue;
```
The code is revamped in [10fabf4](https://github.com/lightningdevkit/rust-lightning/commit/10fabf4266a33bf3ee7684eea54bdf55a328846a)

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4831,6 +4852,14 @@ where
+							let _ = self.force_close_sending_error(&chan_id, &counterparty_node_id, true);
+						}
+					}
+					peer_state.unnotified_outbound_channel.retain(|_, timer| timer.ticks_remaining > 0);
```
The code is revamped in [10fabf4](https://github.com/lightningdevkit/rust-lightning/commit/10fabf4266a33bf3ee7684eea54bdf55a328846a)

shaavan commented:
Thanks, @wpaulino, for the details about the message transmissions!

Seems like it's worth considering extending the PR from fixing the original issue to not failing channel creation mid-handshake due to channel disconnection.

I am tinkering with an approach, and I shall update the PR very soon!

shaavan commented:
**Update:**

Okay, so I have figured out an approach, but this depends on the behavior changes introduced in #2760.

We can track the list of msg_events we have sent during the handshake process, which can be used in case a peer disconnects midway.

Once the funding is signed, we graduate the channel from `OutboundV1` to `Channel`. And so after that, we stop tracking the msg_events.

However, currently, in the main, we graduate the channel as soon as we have created the funding.

```rust
let (chan: Channel<SP>, msg_opt) = match peer_state.channel_by_id.remove(temporary_channel_id) {

...

},
```

Since #2760 is already getting approval and will soon be merged, I shall build this new approach over the changes introduced there.

shaavan commented:
Updated from [pr2725.04](https://github.com/shaavan/rust-lightning/commits/pr2725.04) -> [pr2725.05](https://github.com/shaavan/rust-lightning/commits/pr2725.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.04..pr2725.05))
Addressed @wpaulino and @TheBlueMatt comments

Updates:
1. Rebased on main.
2. **Approach update**

**Logic:**

-> Follow the standard handshake routine.

-> If we disconnect mid-handshake from our peer (that is, OutboundV1Channel is not resolved to a funded channel), we don't immediately close the OutboundV1Channel. 

-> Instead, we track how long it has been since we disconnected from peers.

-> If we connect back within time, we rebroadcast SendOpenChannel corresponding to OutboundV1Channel to the peer.

-> If we do not connect back within N (=2) timer ticks, we force close and remove the channel.

**Note:**
1. To also handle the case of further disconnection mid-handshake, the timer resets when the peer connects back.

shaavan commented:
Updated from [pr2725.05](https://github.com/shaavan/rust-lightning/commits/pr2725.05) -> [pr2725.06](https://github.com/shaavan/rust-lightning/commits/pr2725.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.05..pr2725.06))

Changes:
1. Updated the ClosureReason from `HolderForceClosed` -> `PeerDisconnected` as it is more apt.
2. Updated the test introduced in this PR accordingly.
3. Updated other relevant tests to account for the behavior change introduced in this PR.

benthecarman commented:
Looks like you have a compile issue

shaavan commented on lightning/src/ln/channel.rs:
```diff
@@ -6075,6 +6112,7 @@ impl<SP: Deref> Channel<SP> where
 pub(super) struct OutboundV1Channel<SP: Deref> where SP::Target: SignerProvider {
 	pub context: ChannelContext<SP>,
 	pub unfunded_context: UnfundedChannelContext,
+	pub unconfirmed_context: UnconfirmedChannelContext,
```
I think keeping the `unfunded_context` around is a good idea.

Say we sent an outbound request to a peer, and the peer neither disconnects from us nor sends back an AcceptChannel message.
Then, we need a measure to deal appropriately with such stale OuboundChannel requests.

shaavan replied:
I considered adding it to the `UnfundedChannelContext`. But the timer in that context is shared between Inbound and Outbound, while the introduced here is specific to Outbound only. If we introduce it to both, it might complicate things with peer disconnection handling. So, after giving it some thought, I decided against it.

Let me know what you think about this and if we should go about sharing the timer between both Channel contexts and handling mid-handshake disconnection for the Inbound Channel too.

Looking forward to your thoughts!

jkczyz replied:
@TheBlueMatt Any reason we shouldn't do this for inbound channels, too? IIUC, we'll still be in the unfunded state after sending the `accept_channel` message ourself.

jkczyz replied:
Also, regarding naming, if we decided on a separate struct specific to outbound channels, we should name it in that manner, i.e., `UnfundedOutboundChannelContext`, rather than to one specific (though only) use of it.

jkczyz replied:
Isn't the second timer reset upon reconnection while the original timer is not? We'd at least need a second timer, if not a struct.

jkczyz replied:
Not quite sure I follow what you mean by my reading of #2096. We currently wait `UNFUNDED_CHANNEL_AGE_LIMIT_TICKS` but drop the channel if the peer disconnects. The issue says that if the peer disconnects we should wait one tick before dropping the channel instead. Are you saying to keep the age-based approach as-is but drop the channel if the peer is disconnected on the _next_ timer tick (i.e., not waiting for one more?)

shaavan replied:
Thanks a lot, @TheBlueMatt and @jkczyz, for your thorough discussion

@TheBlueMatt
A final confirmation before going to implementation. Are you suggesting completely dropping the added new timer, and keeping the OutboundV1Channel around for UNFUNDED_CHANNEL_AGE_LIMIT_TICKS even if the peer disconnects (which in the worst case would mean keeping a stale outbound channel request around for 1 hour)?

shaavan commented:
Updated from [pr2725.06](https://github.com/shaavan/rust-lightning/commits/pr2725.06) -> [pr2725.07](https://github.com/shaavan/rust-lightning/commits/pr2725.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.06..pr2725.07))
Addressed @TheBlueMatt comment

Update:

1. Squashed the updates introduced in tests with the changes that broke them so that each commit now individually passes all the tests.

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -919,6 +923,39 @@ impl<'a, SP: Deref> ChannelPhase<SP> where
+	/// created. If this unaccepted channel reaches peer has yet to respond after reaching
+	/// `UNACCEPTED_CHANNEL_AGE_LIMIT_TICKS`, it will be force-closed and purged from memory.
+	///
+	/// This is so that we don't keep outbound request around which have not been accepted
```
s/which/that

shaavan replied:
I have cleaned up this section in [5a4a8bb](https://github.com/lightningdevkit/rust-lightning/pull/2725/commits/5a4a8bb45fa02c39fb9b54ed536e3dc879139348)
Thanks!

shaavan commented:
Updated from [pr2725.07](https://github.com/shaavan/rust-lightning/commits/pr2725.07) -> [pr2725.08](https://github.com/shaavan/rust-lightning/commits/pr2725.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.07..pr2725.08))
Addressed @jkczyz comment

Update:
1. Renamed `unconfirmed_channel_context` -> `disconnected_unfunded_channel_context` to make it better suit its role.
2. Also renamed the unfunded_channel_context -> `stale_unfunded_channel_context` to make it more specific and easier to distinguish from `disconnected_unfunded_channel_context`.
3. Rewrote the comment for the `disconnected_unfunded_channel_context` to more clearly explain its functionality.

shaavan commented on lightning/src/ln/channel.rs:
```diff
@@ -919,6 +923,39 @@ impl<'a, SP: Deref> ChannelPhase<SP> where
+	/// This is so that we don't keep the outbound request around for long whose associated peer
+	/// has disconnected in middle of channel creation handshake, and has not connected back yet.
+	unconfirmed_channel_age_ticks: usize,
+}
```
Thanks for catching it
I agree that it's unwise to overload the name `unconfimed_channel`.
I have renamed the struct, which resolves this problem.
Let me know if you find the new naming more suitable.
Thanks a lot!

shaavan commented on lightning/src/ln/channel.rs:
```diff
@@ -919,6 +923,39 @@ impl<'a, SP: Deref> ChannelPhase<SP> where
+pub(super) struct UnconfirmedChannelContext {
+	/// A counter tracking how many ticks have elapsed since this unaccepted channel was
+	/// created. If this unaccepted channel reaches peer has yet to respond after reaching
+	/// `UNACCEPTED_CHANNEL_AGE_LIMIT_TICKS`, it will be force-closed and purged from memory.
```
Yes!
I was going to remove this comment in the PR cleanup before posting, but this must have slipped from my mind at that time! 😅
Thanks for catching it!

jkczyz commented on lightning/src/ln/channel.rs:
```diff
@@ -919,25 +923,50 @@ impl<'a, SP: Deref> ChannelPhase<SP> where
+	/// the outbound channel will be force-closed and purged from memory. This is so that
+	/// we don't keep the outbound request around for long whose associated peer has
+	/// disconnected in the middle of the channel creation handshake and has not yet connected back.
+	disconnected_unfunded_channel_age_ticks: usize,
```
Reason being "channel age" is not applicable here since this value is reset to zero upon reconnection.

shaavan replied:
I agree with your point!

I expanded upon your suggestion and renamed the variable to `disconnected_peer_channel_ticks`, to make the name more explicit and precise.
Thanks for your suggestion!

jkczyz replied:
I don't think we want to qualify this field with "channel". That is used in the struct name to qualify the context.

shaavan commented:
Updated from [pr2725.08](https://github.com/shaavan/rust-lightning/commits/pr2725.08) -> [pr2725.09](https://github.com/shaavan/rust-lightning/commits/pr2725.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.08..pr2725.09))
Addressed @jkczyz comment

Updates:
- Renamed `disconnected_unfunded_channel_context` -> `disconnected_peer_channel_context`, and made other subsequent name changes. For reviewing the full update, see [diff](https://github.com/shaavan/rust-lightning/compare/pr2725.08..pr2725.09).

@jkczyz

I've taken a look at your name change suggestion, and I like the new names for the variables! I even expanded on your idea a bit, and I think it's a definite improvement from the previous version.

Regarding the introduced timer, I considered adding it to the `UnfundedChannelContext`, but there's a catch. The timer in that context is shared between Inbound and Outbound, while the one in this PR is specific to Outbound only. If we introduce it to both, it might complicate things with peer disconnection handling. So, after giving it some thought, I decided against it.

Let me know what you think about this and if we should go about sharing the timer between both Channel contexts and handling mid-handshake disconnection for the Inbound Channel too.

Looking forward to your thoughts!

shaavan commented on lightning/src/ln/channel.rs:
```diff
@@ -919,25 +923,50 @@ impl<'a, SP: Deref> ChannelPhase<SP> where
 	/// This is so that we don't keep channels around that haven't progressed to a funded state
 	/// in a timely manner.
-	unfunded_channel_age_ticks: usize,
+	stale_unfunded_channel_age_ticks: usize,
```
Yep! Agree with you here!

Addressed in [8ff26c7](https://github.com/lightningdevkit/rust-lightning/commit/8ff26c7072fbdee88c5957e23fce58df778ab71f)

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -11421,6 +11441,89 @@ mod tests {
 		}
 	}
 
+	#[test]
```
Agreed
Let me move this into an appropriate file with the next update
Thanks!

shaavan replied:
Addressed in [59e6af3](https://github.com/lightningdevkit/rust-lightning/commit/59e6af3cc67cd52a3cd27075ff488fce1a3ca3c6) 🚀

shaavan commented:
Updated from [pr2725.09](https://github.com/shaavan/rust-lightning/commits/pr2725.09) -> [pr2725.10](https://github.com/shaavan/rust-lightning/commits/pr2725.10) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.09..pr2725.10))
Addressed @TheBlueMatt comment

- Removed additional introduced timer.
- Instead reverted the code to remove OutboundChannel when it has become stale for UNFUNDED_CHANNEL_AGE_LIMIT_TICKS
- Moved the introduced tests to functional_tests.rs

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10511,6 +10511,90 @@ fn test_remove_expired_inbound_unfunded_channels() {
+		let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+		let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+		// Simulate peer-diconnects mid-handshake
```
disconnects

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10511,6 +10511,90 @@ fn test_remove_expired_inbound_unfunded_channels() {
+
+		// Simulate peer-diconnects mid-handshake
+		// The channel is initiated from the node 0 side,
+		// But the nodes disconnects before node 1 could send accept channel
```
disconnect

shaavan replied:
Fixed in 15792af
Thanks for catching it!

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10511,6 +10511,90 @@ fn test_remove_expired_inbound_unfunded_channels() {
+
+		// Simulate peer-diconnects mid-handshake
+		// The channel is initiated from the node 0 side,
+		// But the nodes disconnects before node 1 could send accept channel
```
s/But/but

shaavan replied:
Fixed in 15792af
Thanks!

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10511,6 +10511,90 @@ fn test_remove_expired_inbound_unfunded_channels() {
+			nodes[0].node.timer_tick_occurred();
+		}
+
+		// Since we disconnected from peer and did not connect back within time
+		// We should have forced-closed the channel by now.
```
nit: "... time, we should have ..."

shaavan replied:
Addressed in 15792af
Thanks!

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10511,6 +10511,90 @@ fn test_remove_expired_inbound_unfunded_channels() {
+		nodes[1].node.peer_disconnected(&nodes[0].node.get_our_node_id());
+
+		{
+			// Make sure that we have not removed the OutboundV1Channel from node[0] immediately.
+			let node_0_per_peer_state = nodes[0].node.per_peer_state.read().unwrap();
+			let per_peer_state = node_0_per_peer_state.get(&nodes[1].node.get_our_node_id()).unwrap().lock().unwrap();
+			assert_eq!(per_peer_state.channel_by_id.len(), 1);
```
Should we also check that `nodes[1]` _does_ drop the peer?

shaavan replied:
I think that's a great idea!
Updated the test in 15792af

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10511,6 +10511,90 @@ fn test_remove_expired_inbound_unfunded_channels() {
+		assert_eq!(events.len(), 1);
+	}
+
+
```
Remove extra blank line

shaavan replied:
Fixed in 15792af
Thanks for catching it!

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10511,6 +10511,90 @@ fn test_remove_expired_inbound_unfunded_channels() {
 }
 
+#[test]
+	fn test_channel_close_when_not_timely_accepted() {
```
Looks like you need to remove one level of indentation.

shaavan commented:
Updated from [pr2725.10](https://github.com/shaavan/rust-lightning/commits/pr2725.10) -> [pr2725.11](https://github.com/shaavan/rust-lightning/commits/pr2725.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.10..pr2725.11))
Addressed @jkczyz and @coderabbitai comments

- Update comments to fix grammatical mistakes and better express the intended behavior
- Stop ignoring `test_peer_disconnected_before_funding_broadcasted`, and instead update it to account for behavior updates.
- Expand the introduced test to check if node[1] has immediately dropped the Unfunded Inbound Channel

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -3692,10 +3692,16 @@ fn test_dup_events_on_peer_disconnect() {
+// The following test is disabled because we no longer close the channel
+// immediately after funding brodcasts. Instead we wait for some time for
+// the peer to reconnect back, and only close it after it become stale for
+// UNFUNDED_CHANNEL_AGE_LIMIT_TICKS
```
Removed the comment
Thanks!

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -3724,8 +3730,8 @@ fn test_peer_disconnected_before_funding_broadcasted() {
-	// Ensure that the channel is closed with `ClosureReason::DisconnectedPeer` when the peers are
-	// disconnected before the funding transaction was broadcasted.
+	// Ensure that the channel is closed after timeout with `ClosureReason::DisconnectedPeer`
+	// when the peers are disconnected before the funding transaction was broadcasted.
```
Updated the comment
Thanks!

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -893,6 +893,7 @@ impl <SP: Deref> PeerState<SP> where SP::Target: SignerProvider {
 			return false
 		}
 		self.channel_by_id.iter().filter(|(_, phase)| matches!(phase, ChannelPhase::Funded(_))).count() == 0
+			&& self.channel_by_id.iter().filter(|(_, phase)| matches!(phase, ChannelPhase::UnfundedOutboundV1(_))).count() == 0
```
Thanks for the suggestion!
Addressed in 15792af

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9011,15 +9014,22 @@ where
+							msg: chan.get_open_channel(self.chain_hash),
+						});
+					}
+					// else don't do anything if the channel is UnfundedInbound Channel.
```
Update the code in 15792af
Thanks for the suggestion!

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10688,6 +10694,11 @@ fn test_disconnect_in_funding_batch() {
+		nodes[0].node.timer_tick_occurred();
+	}
+
 	// The channels in the batch will close immediately.
```
Addressed in 15792af
Thanks!

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -3692,10 +3692,16 @@ fn test_dup_events_on_peer_disconnect() {
 fn test_peer_disconnected_before_funding_broadcasted() {
 	// Test that channels are closed with `ClosureReason::DisconnectedPeer` if the peer disconnects
-	// before the funding transaction has been broadcasted.
+	// before the funding transaction has been broadcasted, and doesn't reconnect back within time.
```
Updated the test to account for new behavior instead of ignoring it!
Thanks for the suggestion!

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -3724,12 +3724,17 @@ fn test_peer_disconnected_before_funding_broadcasted() {
 
-	// Ensure that the channel is closed with `ClosureReason::DisconnectedPeer` when the peers are
-	// disconnected before the funding transaction was broadcasted.
+	// Ensure that the channel is closed if the peer does not reconnect within the timeout period
+	// when the peers are disconnected before the funding transaction was broadcasted.
```
This is a bit of a mouthful now. Instead of modifying the comment, let's move it above `check_closed_event`.

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10688,7 +10789,12 @@ fn test_disconnect_in_funding_batch() {
 	nodes[0].node.peer_disconnected(&nodes[2].node.get_our_node_id());
 
-	// The channels in the batch will close immediately.
+	// After the time expires for allowing peer to connect back
```
This reads a bit incomplete. Let's rephrase this in similar comments elsewhere as:

```rust
// Let the time for the peer to reconnect expire.
```

shaavan replied:
Thanks for the suggestion!
Addressed it in [f9fbdca](https://github.com/lightningdevkit/rust-lightning/commit/f9fbdcad5e018c8533fc3670bba3dd4b9fd281ce)

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9011,15 +9015,31 @@ where
+
+						ChannelPhase::UnfundedInboundV1(_) => {
+							// Since unfunded inbound channel maps are cleared upon disconnecting a peer, and they're not
+							// persisted (so won't be recovered after a crash), they shouldn't exist here and we would never
+							// need to worry about closing and removing them.
```
This sentence is a bit lengthy, and the second portion isn't accurate as we aren't doing clean-up here.

Let's re-phrase and tighten this up a bit as:

Since unfunded inbound channel maps are cleared upon disconnecting a peer, they're not persisted so won't be recovered after a crash. Therefore, they shouldn't exist here.

shaavan replied:
Addressed it in [f9fbdca](https://github.com/lightningdevkit/rust-lightning/commit/f9fbdcad5e018c8533fc3670bba3dd4b9fd281ce)
Thanks for the suggestion on rephrasing!

jkczyz reviewed (commented):
Looks good other than some comment re-phrasing.

shaavan commented:
Updated from [pr2725.11](https://github.com/shaavan/rust-lightning/commits/pr2725.11) -> [pr2725.12](https://github.com/shaavan/rust-lightning/commits/pr2725.12) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.11..pr2725.12))
Addressed @jkczyz comments

- Improve the comment messages.

shaavan commented:
@TheBlueMatt
A gentle ping.
I think this PR is ready for a (potentially final) review :)

shaavan commented:
Updated from [pr2725.12](https://github.com/shaavan/rust-lightning/commits/pr2725.12) -> [pr2725.13](https://github.com/shaavan/rust-lightning/commits/pr2725.13) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.12..pr2725.13))
Addressed @TheBlueMatt comments

1. Repurpose the `test_disconnect_in_funding_batch` to check if all the channels of the batch close if one of them is closed.
2. Use `nodes[0].node.list_channels()` in introduced test for cleaner code.
3. Made the added test more precise by adding a test for correct msg-type along with the number of msgs.

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10688,7 +10791,12 @@ fn test_disconnect_in_funding_batch() {
 
-	// The channels in the batch will close immediately.
+	// Let the time for the peers to reconnect expire.
+	for _ in 0..UNFUNDED_CHANNEL_AGE_LIMIT_TICKS {
```
I agree with your point.
Addressed this in 0fd3d31ec6f6cddb3d3363cd2c7b834f0a858373

Thanks!

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10505,6 +10512,102 @@ fn test_remove_expired_inbound_unfunded_channels() {
+	{
+		// Make sure that we have not removed the OutboundV1Channel from node[0] immediately.
+		let node_0_per_peer_state = nodes[0].node.per_peer_state.read().unwrap();
+		let per_peer_state = node_0_per_peer_state.get(&nodes[1].node.get_our_node_id()).unwrap().lock().unwrap();
```
Addressed in 0fd3d31ec6f6cddb3d3363cd2c7b834f0a858373
Thanks for the tip!

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10505,6 +10512,102 @@ fn test_remove_expired_inbound_unfunded_channels() {
+	// Make sure the SendOpenChannel message is added to
+	// node_0 pending message events
+	let events = nodes[0].node.get_and_clear_pending_msg_events();
+	assert_eq!(events.len(), 1);
```
Added the check of specific msg-type in 0fd3d31ec6f6cddb3d3363cd2c7b834f0a858373
Thanks!

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -3695,7 +3695,7 @@ fn test_dup_events_on_peer_disconnect() {
 fn test_peer_disconnected_before_funding_broadcasted() {
 	// Test that channels are closed with `ClosureReason::DisconnectedPeer` if the peer disconnects
-	// before the funding transaction has been broadcasted.
+	// before the funding transaction has been broadcasted, and doesn't reconnect back within time.
```
This test is already present in the main branch.
Addressing the changes in this PR, I have made appropriate modifications to this test.

shaavan commented:
Updated from [pr2725.13](https://github.com/shaavan/rust-lightning/commits/pr2725.13) -> [pr2725.14](https://github.com/shaavan/rust-lightning/commits/pr2725.14) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2725.13..pr2725.14))
Addressed @TheBlueMatt comments
- Use `list_channels` in the introduced test wherever necessary.
- Made the check for `msg_events` more thorough by checking for the specific expected message.
- Clean up commit structure.

Updated from [pr2725.14](https://github.com/shaavan/rust-lightning/commits/pr2725.14) -> [pr2725.15](https://github.com/shaavan/rust-lightning/commits/pr2725.15)

<details>
<summary>
Range-diff
</summary>

```
git range-diff fbeb7ac9e1256f5d69b0cda580b56c5025ffa987..411462cce2c23d5a3faa72641107f4d2e43cf5a5 5bf58f0d33b13bd3d9f88c4f2021fbd77745274a..ddf75afd167ac4adb5824180b4125d87563fa31a
1:  411462cc ! 1:  ddf75afd Do not remove Outbound Channel immediately when peer disconnects
    @@ lightning/src/ln/functional_tests.rs: fn test_disconnect_in_funding_batch() {
     +  // Force-close the channel for which we've completed the initial monitor.
        let funding_txo_1 = OutPoint { txid: tx.txid(), index: 0 };
        let funding_txo_2 = OutPoint { txid: tx.txid(), index: 1 };
    -   let channel_id_1 = funding_txo_1.to_channel_id();
    -   let channel_id_2 = funding_txo_2.to_channel_id();
    +   let channel_id_1 = ChannelId::v1_from_funding_outpoint(funding_txo_1);
    +   let channel_id_2 = ChannelId::v1_from_funding_outpoint(funding_txo_2);
     +
     +  nodes[0].node.force_close_broadcasting_latest_txn(&channel_id_1, &nodes[1].node.get_our_node_id()).unwrap();
     +
```
</details>

- Rebased over main

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10505,6 +10512,97 @@ fn test_remove_expired_inbound_unfunded_channels() {
+		// Since channel was inbound from node[1] perspective, it should have been immediately dropped.
+		let node_1_per_peer_state = nodes[1].node.per_peer_state.read().unwrap();
+		let per_peer_state = node_1_per_peer_state.get(&nodes[0].node.get_our_node_id());
+		assert!(per_peer_state.is_none());
```
Addressed in [a6a6b48](https://github.com/lightningdevkit/rust-lightning/commit/a6a6b481020cf3c3cc9591b5c0306fbbc7270fec)
Thanks!

shaavan commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10686,14 +10779,39 @@ fn test_disconnect_in_funding_batch() {
+		_ => panic!("Unexpected message."),
+	}
+
+	// We broadcast the commitment transaction as part of the force-close.
```
Super interested in understanding the issue here! And probably might give it a try if it's not a super biggie!
